azure/compute/src/main/java/org/jclouds/azurecompute/compute/functions/VirtualMachineToNodeMetadata.java

```diff
@@ -0,0 +1,109 @@
+   private final Supplier<Set<? extends Hardware>> hardwares;
+   private final Supplier<Set<? extends Location>> locations;
+   private final GroupNamingConvention nodeNamingConvention;
+   private final Map<DeploymentStatus, Status> deploymentStatusToStatus;
+   private final Map<String, Status> instanceStatusToStatus;
+
+   private static final Function<VirtualIPAddress, String> toIpAddress = new Function<VirtualIPAddress, String>() {
+       @Override
+       public String apply(VirtualIPAddress input) {
+           return input.getAddress() == null ? null : input.getAddress().getHostAddress();
+       }
+   };
+
+   @Inject
+   VirtualMachineToNodeMetadata(@Memoized Supplier<Set<? extends Image>> images, @Memoized Supplier<Set<? extends Hardware>> hardwares,
+                                Supplier<Set<? extends Location>> locations,
+                                GroupNamingConvention.Factory namingConvention,
+                                Map<DeploymentStatus, Status> deploymentStatusToStatus, Map<String, Status> instanceStatusToStatus) {
+      this.nodeNamingConvention = checkNotNull(namingConvention, "namingConvention").createWithoutPrefix();
+      this.images = checkNotNull(images, "images");
+      this.hardwares = checkNotNull(hardwares, "hardwares");
+      this.locations = checkNotNull(locations, "locations");
+      this.deploymentStatusToStatus = checkNotNull(deploymentStatusToStatus);
+      this.instanceStatusToStatus = instanceStatusToStatus;
+   }
+
+   @Override
+   public NodeMetadata apply(VirtualMachine from) {
+      NodeMetadataBuilder builder = new NodeMetadataBuilder();
+      builder.ids(from.getServiceName() + ":" + from.getRoleName());
+      builder.name(from.getServiceName() + ":" + from.getRoleName());
+//      Location location = Iterables.find(locations.get(), LocationPredicates.idEquals(from.getDatacenter().getId() + ""));
+//      builder.location(location);
+//      builder.group(nodeNamingConvention.groupInUniqueNameOrNull(from.getname()));
+      builder.group(from.getServiceName());
+      if (from.getRole().getRoleSize() != null) {
+         builder.hardware(from(hardwares.get()).firstMatch(HardwarePredicates.idEquals(from.getRole().getRoleSize())).orNull());
+      }
+      if (from.getRole().getOSVirtualHardDisk() != null) {
+         String imageId = from.getRole().getOSVirtualHardDisk().getSourceImageName();
+         builder.imageId(imageId);
+         Image image = from(images.get()).firstMatch(ImagePredicates.idEquals(imageId)).orNull();
+         if (image != null) {
+             builder.operatingSystem(image.getOperatingSystem());
+         }
+      }
+      builder.status(deploymentStatusToStatus.get(from.getDeploymentStatus()));
+
+//      builder.status(serverStateToNodeStatus.get(from.getState()));
+      builder.publicAddresses(ImmutableList.copyOf(Collections2.transform(from.getPublicIpAddresses(), toIpAddress)));
+      return builder.build();
```

Comment:
The node should also try to populate the credentials. If the API does not provide them, it is usually a good idea to read the credentials from the jclouds credential store as done in [DigitalOcean](https://github.com/jclouds/jclouds-labs/blob/master/digitalocean/src/main/java/org/jclouds/digitalocean/compute/functions/DropletToNodeMetadata.java#L115-L120).